core/snippets/snippets_parser.py

```diff
@@ -0,0 +1,339 @@
+import re
+from pathlib import Path
+from typing import Callable, Union
+
+from .snippet_types import Snippet, SnippetVariable
+
+
+class SnippetDocument:
+    file: str
+    line_doc: int
+    line_body: int
+    variables: list[SnippetVariable] = []
+    name: str = None
+    phrases: list[str] = None
+    insertionScopes: list[str] = None
+    languages: list[str] = None
+    body: str = None
+
+    def __init__(self, file: str, line_doc: int, line_body: int):
+        self.file = file
+        self.line_doc = line_doc
+        self.line_body = line_body
+
+
+def create_snippets_from_file(file_path: str) -> list[Snippet]:
+    documents = parse_file(file_path)
+    return create_snippets(documents)
+
+
+def create_snippets(documents: list[SnippetDocument]) -> list[Snippet]:
+    if len(documents) == 0:
+        return []
+
+    if documents[0].body is None:
+        default_context = documents[0]
+        documents = documents[1:]
+    else:
+        default_context = SnippetDocument("", -1, -1)
+
+    snippets: list[Snippet] = []
+
+    for doc in documents:
+        snippet = create_snippet(doc, default_context)
+        if snippet:
+            snippets.append(snippet)
+
+    return snippets
+
+
+def create_snippet(
+    document: SnippetDocument, default_context: SnippetDocument
+) -> Snippet:
+    snippet = Snippet(
+        name=document.name or default_context.name,
+        languages=document.languages or default_context.languages,
+        phrases=document.phrases or default_context.phrases,
+        variables=combine_variables(default_context.variables, document.variables),
+        body=normalize_snippet_body_tabs(document.body),
+    )
+
+    if not validate_snippet(document, snippet):
+        return None
+
+    return snippet
+
+
+def validate_snippet(document: SnippetDocument, snippet: Snippet) -> bool:
```

Comment:
Not at all
`substitutions[variable.name] = formatted_phrase` is scoped on the variable name so there is no problem with multiple variables being substituted.

`insertions_phrase[phrase] = id_ctx` just points the phrase to the snippet. Every variable in the same snippet will have the same `id_ctx`

Reply:
Oh ok I see. So if you have two variables with insertion formatter, it will just insert the phrase into each one with a different formatter? That's cool

Reply:
I thought so :D